bpf/lib/nodeport.h

```diff
@@ -1464,8 +1597,32 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
 #endif /* DSR_ENCAP_MODE */
 
 static __always_inline int
-handle_dsr_v4(struct __ctx_buff *ctx, struct iphdr *ip4, bool *dsr)
+nodeport_extract_dsr_v4(struct __ctx_buff *ctx, struct iphdr *ip4,
+			struct ipv4_ct_tuple *tuple, int l4_off, __be32 *addr,
+			__be16 *port, bool *dsr)
 {
+	struct ipv4_ct_tuple tmp = *tuple;
+
+	/* TCP connections have the DSR Option only in their SYN packet: */
+	if (tuple->nexthdr == IPPROTO_TCP) {
+		union tcp_flags tcp_flags = {};
+
+		if (l4_load_tcp_flags(ctx, l4_off, &tcp_flags) < 0)
+			return DROP_CT_INVALID_HDR;
+
+		ipv4_ct_tuple_reverse(&tmp);
+
+		if (!(tcp_flags.value & TCP_FLAG_SYN)) {
+			/* Packet belongs to a tracked DSR connection, so
```

Comment:
I'm a bit puzzled how you decide that a `packet belongs to a tracked DSR connection`?  I think this branch can be hit not only by LB traffic.

Reply:
The `ct_has_dsr_egress_entry4()` lookup below tells us if a preceding SYN for this TCP connection created a DSR CT entry. So maybe the comment would be better phrased as

> If the packet belongs to a tracked DSR connection, trigger a CT update:

?

Reply:
Yep, the suggestion :+1:

Reply:
added